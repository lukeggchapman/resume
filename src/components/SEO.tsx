/**
 * SEO component that queries for data with
 *  Gatsby's useStaticQuery React hook
 *
 * See: https://www.gatsbyjs.org/docs/use-static-query/
 */

import React from 'react'
import PropTypes from 'prop-types'
import Helmet, { HelmetProps } from 'react-helmet'
import { useStaticQuery, graphql } from 'gatsby'

export interface SEOProps {
  title: string
  description: string
  lang: string
  keywords: string[]
  meta: HelmetProps['meta']
}

const SEO = ({ description, lang, meta, keywords, title }: SEOProps) => {
  const { site } = useStaticQuery(
    graphql`
      query {
        site {
          siteMetadata {
            title
            description
            author
          }
        }
      }
    `
  )

  const metaDescription = description || site.siteMetadata.description

  const test: HelmetProps['meta'] = [
    {
      name: `description`,
      content: metaDescription,
    },
    {
      property: `og:title`,
      content: title,
    },
    {
      property: `og:description`,
      content: metaDescription,
    },
    {
      property: `og:type`,
      content: `website`,
    },
    {
      name: `twitter:card`,
      content: `summary`,
    },
    {
      name: `twitter:creator`,
      content: site.siteMetadata.author,
    },
    {
      name: `twitter:title`,
      content: title,
    },
    {
      name: `twitter:description`,
      content: metaDescription,
    },
  ]

  return (
    <Helmet
      htmlAttributes={{
        lang,
      }}
      title={title}
      titleTemplate={`%s | ${site.siteMetadata.title}`}
      meta={test
        .concat(
          keywords.length > 0
            ? {
                name: `keywords`,
                content: keywords.join(`, `),
              }
            : []
        )
        .concat(meta || [])}
    />
  )
}

SEO.defaultProps = {
  lang: `en`,
  meta: [],
  keywords: [],
  description: ``,
}

SEO.propTypes = {
  description: PropTypes.string,
  lang: PropTypes.string,
  meta: PropTypes.arrayOf(PropTypes.object),
  keywords: PropTypes.arrayOf(PropTypes.string),
  title: PropTypes.string.isRequired,
}

export default SEO
