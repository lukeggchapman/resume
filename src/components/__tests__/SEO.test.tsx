import React from 'react'
import { shallow } from 'enzyme'
import { useStaticQuery } from 'gatsby'

import SEO, { SEOProps } from '../SEO'

jest.mock('react-helmet', () => 'Helmet')

const useStaticQueryMock = useStaticQuery as jest.Mock

const baseProps: SEOProps = {
  title: 'titleTest',
  description: 'descriptionTest',
  lang: 'langTest',
  keywords: ['keywordTest1', 'keywordTest1'],
  meta: [
    {
      name: 'extendedMeta',
      content: 'Test',
    },
  ],
}

describe('SEO', () => {
  useStaticQueryMock.mockReturnValue({
    site: {
      siteMetadata: {
        description: 'siteDescriptionTest',
        author: 'authorDescriptionTest',
        title: 'titleTest',
      },
    },
  })

  const render = () => {
    return shallow(<SEO {...baseProps} />)
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('renders correctly', () => {
    expect(render()).toMatchSnapshot()
  })

  it('renders with defaultProps', () => {
    expect(shallow(<SEO title="defaultsTestTitle" />)).toMatchSnapshot()
  })

  it('graphql query remains unchanged', () => {
    render()

    expect(useStaticQueryMock).toHaveBeenCalledTimes(1)
    expect(useStaticQueryMock.mock.calls[0][0]).toMatchSnapshot()
  })
})
