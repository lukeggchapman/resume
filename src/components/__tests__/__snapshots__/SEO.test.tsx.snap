// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SEO graphql query remains unchanged 1`] = `
Array [
  "
      query {
        site {
          siteMetadata {
            title
            description
            author
          }
        }
      }
    ",
]
`;

exports[`SEO renders correctly 1`] = `
<Helmet
  htmlAttributes={
    Object {
      "lang": "langTest",
    }
  }
  meta={
    Array [
      Object {
        "content": "descriptionTest",
        "name": "description",
      },
      Object {
        "content": "titleTest",
        "property": "og:title",
      },
      Object {
        "content": "descriptionTest",
        "property": "og:description",
      },
      Object {
        "content": "website",
        "property": "og:type",
      },
      Object {
        "content": "summary",
        "name": "twitter:card",
      },
      Object {
        "content": "authorDescriptionTest",
        "name": "twitter:creator",
      },
      Object {
        "content": "titleTest",
        "name": "twitter:title",
      },
      Object {
        "content": "descriptionTest",
        "name": "twitter:description",
      },
      Object {
        "content": "keywordTest1, keywordTest1",
        "name": "keywords",
      },
      Object {
        "content": "Test",
        "name": "extendedMeta",
      },
    ]
  }
  title="titleTest"
  titleTemplate="%s | titleTest"
/>
`;

exports[`SEO renders with defaultProps 1`] = `
<Helmet
  htmlAttributes={
    Object {
      "lang": "en",
    }
  }
  meta={
    Array [
      Object {
        "content": "siteDescriptionTest",
        "name": "description",
      },
      Object {
        "content": "defaultsTestTitle",
        "property": "og:title",
      },
      Object {
        "content": "siteDescriptionTest",
        "property": "og:description",
      },
      Object {
        "content": "website",
        "property": "og:type",
      },
      Object {
        "content": "summary",
        "name": "twitter:card",
      },
      Object {
        "content": "authorDescriptionTest",
        "name": "twitter:creator",
      },
      Object {
        "content": "defaultsTestTitle",
        "name": "twitter:title",
      },
      Object {
        "content": "siteDescriptionTest",
        "name": "twitter:description",
      },
    ]
  }
  title="defaultsTestTitle"
  titleTemplate="%s | titleTest"
/>
`;
